### ISSUE TITLE
Crash with: Unhandled Exception: System.ArgumentNullException: A runtime is required. Arg_ParamName_Name, runtime

### ISSUE THREAD

**Issue by james-hu** [microsoft/node-api-dotnet#203]
My use case is calling .NET from Node.
I made it work with Node.js 18.

However, after switching Node.js version to 20, it crached:

```
Unhandled Exception: System.ArgumentNullException: A runtime is required. Arg_ParamName_Name, runtime
   at Microsoft.JavaScript.NodeApi.JSValueScope..ctor(JSValueScopeType, JSRuntime.napi_env, JSRuntime, JSSynchronizationContext) + 0x475
   at Microsoft.JavaScript.NodeApi.JSValue.InvokeCallback[TDescriptor](JSRuntime.napi_env, JSRuntime.napi_callback_info, JSValueScopeType, Func`2) + 0x88
   at Microsoft.JavaScript.NodeApi.JSValue.InvokeJSGetterNoContext(JSRuntime.napi_env, JSRuntime.napi_callback_info) + 0x67
Aborted (core dumped)
```

BTW, the reason I need to switch to Node.js 20 is because my runtime environment with Node.js 18 has a low GLIBC version. This is the error stack:
```
"stack": [
        "Error: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /var/task/node_modules/node-api-dotnet/linux-x64/Microsoft.JavaScript.NodeApi.node)",
        "    at Module._extensions..node (node:internal/modules/cjs/loader:1452:18)",
        "    at Module.load (node:internal/modules/cjs/loader:1197:32)",
        "    at Module._load (node:internal/modules/cjs/loader:1013:12)",
        "    at Module.require (node:internal/modules/cjs/loader:1225:19)",
        "    at require (node:internal/modules/helpers:177:18)",
        "    at initialize (/var/task/node_modules/node-api-dotnet/init.js:48:22)",
        "    at Object.<anonymous> (/var/task/node_modules/node-api-dotnet/index.js:2:18)",
        "    at Module._compile (node:internal/modules/cjs/loader:1356:14)",
        "    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)",
        "    at Module.load (node:internal/modules/cjs/loader:1197:32)"
    ]
```
I am using it in AWS Lambda. Node.js 18 comes with GLIBC 2.26. Node.js 20 comes with GLIBC 2.34.

I will be unblocked by solving any of those two issues.

**Comment by jasongin** [comment]
I don't think we're able to change the glibc requirement, since that comes from .NET Native AOT build tools. However I found this SO question where someone was able to work around that issue in AWS lambda:
https://stackoverflow.com/questions/76232546/version-glibc-2-29-not-found-on-lambda-invocation-using-aot-and-net7

I've been using Node.js v20 for a while for daily development (mostly on Windows). Just now I created PR #204 that includes Node.js v20 in the build & test matrix, and the tests all passed on Linux. So I'm not sure what is the cause of that ArgumentNullException.

What does your JS code look like that loads `node-api-dotnet`?
To collect some additional diagnostic logging, you can set the environment variable: `NODE_API_TRACE_HOST=1`.

**Comment by james-hu** [comment]
Thank you for the suggestions.

I used NODE_API_TRACE_HOST=1 and saw it tyring to initialise with .NET 8.0 which is not installed.

Then I specified net6.0 in the import statement

```
import dotnet from 'node-api-dotnet/net6.0';
``` 

After making that change, the problem is gone.

**Comment by james-hu** [comment]
Sorry I have to reopen this one. I am still trying to figure out what happened but I am seeing a weird situation:

- things work well locally (works on my computer)
- but, in a CI/CD pipeline, it fails with `Unhandled Exception: System.ArgumentNullException: A runtime is required. Arg_ParamName_Name, runtime`

With `NODE_API_TRACE_HOST=1`, I can see that it used net6.0 a expected:

```
> NativeHost.InitializeModule(7F773BE13110, 06B343E0)
< NativeHost.InitializeModule()
> NativeHost.InitializeManagedHost(net6.0, /builds/ai-media/andromeda-squad/formatter/service/node_modules/node-api-dotnet/net6.0/Microsoft.JavaScript.NodeApi.DotNetHost.dll)
    Initializing .NET 6.0
> NativeHost.InitializeManagedRuntime(/builds/ai-media/andromeda-squad/formatter/service/node_modules/node-api-dotnet/net6.0/Microsoft.JavaScript.NodeApi.runtimeconfig.json)
> HostFxr.Initialize()
    .NET root: /usr/share/dotnet
    HostFxr path: /usr/share/dotnet/host/fxr/6.0.26/libhostfxr.so
< HostFxr.Initialize()
    Initializing runtime...
< NativeHost.InitializeManagedRuntime()
    Getting runtime load-assembly delegate...
    Loading managed host type: Microsoft.JavaScript.NodeApi.DotNetHost.ManagedHost, Microsoft.JavaScript.NodeApi.DotNetHost, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
    Invoking managed host method: InitializeModule
> ManagedHost.InitializeModule(7F773BE13110)
    .NET Runtime version: 6.0.26
> ManagedHost.LoadAssemblyTypes(System.Private.CoreLib)
```

, and it loaded DLLs successfully.

Is the "runtime" refered in "A runtime is required" the Javascript/Node runtime? How is the runtime loaded/discovered? Does it depend on a environment variable, or something on the path, or something else?

The situation could be complicated becase we are using nvm to switch Node.js versions. (However, nvm is used in both my laptop and the CI/CD container).

Now I think my previous statement about fixing the issue through `'node-api-dotnet/net6.0'` may not be true.

**Comment by james-hu** [comment]
It would be very helpful if there could be some trace/debug information regarding how JSRuntime (I suppose it is JSRuntime) is created/discovered.

**Comment by james-hu** [comment]
I think I have some clue now.

- The `Unhandled Exception: System.ArgumentNullException: A runtime is required. Arg_ParamName_Name, runtime` error didn't happy before Node.js code calling .NET code. It happened before the code/process/program existed
- So that it seems to be a leaking thing.

It probably is also related to CPU speed, that's why it didn't happy on my laptop but happens in the pipeline in GitLab.

I am now able to reproduce it on my laptop with `npx jest --no-coverage --forceExit --detectOpenHandles`

**Comment by james-hu** [comment]
It seems to be this:

```
Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  JSSynchronizationContext

      3 | import fs from 'node:fs';
      4 | import path from 'node:path';
    > 5 | import dotnet from 'node-api-dotnet';
        | ^
      6 |
      7 | function findDll(name: string, searchPaths: string[]): string | undefined {
      8 |   for (const p of searchPaths) {

      at initialize (node_modules/node-api-dotnet/init.js:49:23)
      at Object.<anonymous> (node_modules/node-api-dotnet/index.js:2:18)
```

Line 49 of init.js is:
```
dotnet = nativeHost.initialize(targetFramework, managedHostPath, require);
```

Is there a processing running in the background? How do we tell that process to quit?

**Comment by james-hu** [comment]
More findings:

- Yes, there's a hidden `dispose()` function on `dotnet` instance
- Calling that dispose function can trigger the core dump

In the code below, the core dump happens after `console.error(dotnet);` but before `console.error('333333333333333');`:
```javascript
    console.error(dotnet);
    try {
      (dotnet as any).dispose();
    } catch(error) {
      console.error(error);
    }
    console.error('333333333333333');
```
Since it is a core dump, `console.error(error);` does not have chance to be executed.

**Comment by james-hu** [comment]
The unbelievable thing is, the problem went away after I chagned the GitLab runner from 2vCPU/8GBmem to 4vCPU/16GB mem :-)

But still, I think the foundamental issue still needs to be addressed. 

Maybe a quick fix could be to add try/catch in the `Dispose()` method in `NativeHost.cs` and print out a warning message?

**Comment by jasongin** [comment]
> The Unhandled Exception: System.ArgumentNullException: A runtime is required. Arg_ParamName_Name, runtime error didn't happy before Node.js code calling .NET code. It happened before the code/process/program existed

The native host error on application shutdown is most likely #216. The error message is different from the one I reported in that bug because another bug caused the ArgumentNullException when attempting to throw a JS Error in that state. But that one was recently fixed by #215.

> It would be very helpful if there could be some trace/debug information regarding how JSRuntime (I suppose it is JSRuntime) is created/discovered.

There is: the `NODE_API_TRACE_HOST` env variable is documented at [README-DEV.md#tracing](https://github.com/microsoft/node-api-dotnet/blob/main/README-DEV.md#tracing)

**Comment by jasongin** [comment]
Closing this as a duplicate of #216. With the latest version of the library I think you'll see a similar stack trace to the one in that issue.
